sound/soc/sof/ipc4-pcm.c

```diff
@@ -191,6 +191,78 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
  * prepare ioctl before the START trigger.
  */
 
+static int sof_ipc4_chain_dma_send_msg(struct snd_sof_dev *sdev,
+				       struct snd_sof_pcm_stream_pipeline_list *pipeline_list,
+				       bool allocate, bool enable)
+{
+	struct sof_ipc4_msg msg = {{ 0 }};
+	int i;
+
+	msg.primary = SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_GLB_CHAIN_DMA);
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
+
+	/*
+	 * To set-up the DMA chain, the host DMA ID and SCS setting
+	 * are retrieved fromthe host pipeline configuration. Likewise
+	 * the link DMA ID and fifo_size are retrieved from the link
+	 * pipeline configuration.
+	 */
+	for (i = 0; i < pipeline_list->count; i++) {
+		struct snd_sof_pipeline *spipe = pipeline_list->pipelines[i];
+		struct snd_sof_widget *pipe_widget = spipe->pipe_widget;
+		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+		if (!pipeline->use_chain_dma) {
+			dev_err(sdev->dev,
+				"All pipelines in chained DMA stream should have use_chain_dma attribute set.");
+			return -EINVAL;
+		}
+
+		msg.primary |= pipeline->msg.primary;
+
+		if (allocate)
+			msg.extension |= pipeline->msg.extension;
+	}
+
+	if (allocate)
+		msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_ALLOCATE_MASK;
+
+	if (enable)
+		msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_ENABLE_MASK;
+
+	return sof_ipc_tx_message(sdev->ipc, &msg, 0, NULL, 0);
+}
+
+/*
+ * Chained DMA is a special case where there is no processing on
+ * DSP. The samples are just moved over host side DMA to a single
+ * buffer on DSP and directly from the to link DMA.  However, the
+ * model on SOF driver has two dummy pipeline, one at host DAI, and
```

Comment:
while you're still fixing typos, here's another one "two dummy pipelines" and we've been asked to use inclusive terms. so maybe two pretend pipelines?

Reply:
I love using "notional", it makes me sound smart.

Reply:
/me checks exact meaning of notional

Seems perfect, and I learned a new word in English :). And fixed one other typo that I found after reading the text for the second time.